ceph-windows-image-build/build/build

```diff
@@ -0,0 +1,173 @@
+rmdir ${BUILD_DIR}/floppy
+
+echo "Starting libvirt VM"
+qemu-img create -f qcow2 ${BUILD_DIR}/ceph-win-ltsc2019-ci-image.qcow2 50G
+VM_NAME="ceph-win-ltsc2019"
+sudo virt-install \
+    --name $VM_NAME \
+    --os-variant win2k19 \
+    --boot hd,cdrom \
+    --virt-type kvm \
+    --graphics spice \
+    --cpu host \
+    --vcpus 4 \
+    --memory 4096 \
+    --disk ${BUILD_DIR}/floppy.img,device=floppy \
+    --disk ${BUILD_DIR}/ceph-win-ltsc2019-ci-image.qcow2,bus=virtio \
+    --disk ${BUILD_DIR}/windows-server-2019.iso,device=cdrom \
+    --disk ${BUILD_DIR}/virtio-win.iso,device=cdrom \
+    --network network=default,model=virtio \
+    --controller type=virtio-serial \
+    --channel unix,target_type=virtio,name=org.qemu.guest_agent.0 \
+    --noautoconsol
+
+export SSH_USER="administrator"
+export SSH_KNOWN_HOSTS_FILE="${BUILD_DIR}/known_hosts"
+export SSH_KEY="$SSH_PRIVATE_KEY"
+
+SECONDS=0
+TIMEOUT=1200
+SLEEP_SECS=30
+while true; do
+    if [[ $SECONDS -gt $TIMEOUT ]]; then
+        echo "Timeout waiting for the VM to start"
+        exit 1
+    fi
+    VM_IP=$(sudo virsh domifaddr --source agent --interface Ethernet --full $VM_NAME | grep ipv4 | awk '{print $4}' | cut -d '/' -f1) || {
+        echo "Retrying in $SLEEP_SECS seconds"
+        sleep $SLEEP_SECS
+        continue
+    }
+    ssh-keyscan -H $VM_IP &> $SSH_KNOWN_HOSTS_FILE || {
+        echo "SSH is not reachable yet"
+        sleep $SLEEP_SECS
+        continue
+    }
+    SSH_ADDRESS=$VM_IP ssh_exec hostname || {
+        echo "Cannot execute SSH commands yet"
+        sleep $SLEEP_SECS
+        continue
+    }
+    break
+done
+export SSH_ADDRESS=$VM_IP
+
+scp_upload ${BUILD_DIR}/install-windows-updates.ps1 /install-windows-updates.ps1
+SSH_TIMEOUT=1h ssh_exec powershell.exe -File /install-windows-updates.ps1
+ssh_exec powershell.exe Remove-Item -Force /install-windows-updates.ps1
+
+restart_windows_vm 1800
+
+scp_upload ${BUILD_DIR}/setup.ps1 /setup.ps1
+SSH_TIMEOUT=1h ssh_exec powershell.exe -File /setup.ps1
+ssh_exec powershell.exe Remove-Item -Force /setup.ps1
+
+restart_windows_vm
+
+sudo virsh qemu-agent-command $VM_NAME \
+    '{"execute":"guest-exec", "arguments":{"path":"ipconfig.exe", "arg":["/release"]}}'
+sudo virsh qemu-agent-command $VM_NAME \
+    '{"execute":"guest-exec", "arguments":{"path":"C:\\Windows\\System32\\Sysprep\\Sysprep.exe", "arg":["/generalize", "/oobe", "/shutdown", "/quiet"]}}'
+
+SECONDS=0
+TIMEOUT=600
+while true; do
+    if [[ $SECONDS -gt $TIMEOUT ]]; then
+        echo "Timeout waiting for the sysprep to shut down the VM"
+        exit 1
+    fi
+    if sudo virsh list | grep -q " $VM_NAME "; then
+        echo "Sysprep is still running"
+        sleep 10
+        continue
+    fi
+    echo "Sysprep finished"
+    break
+done
+
+sudo mv "${BUILD_DIR}/ceph-win-ltsc2019-ci-image.qcow2" ${WORKSPACE}/
+
+${BUILD_DIR}/cleanup
+
+echo "Image successfully built at: ${WORKSPACE}/ceph-win-ltsc2019-ci-image.qcow2"
```

Comment:
@djgalloway

It would be awesome if I can upload this build output directly to:
https://filedump.ceph.com/windows/ceph-win-ltsc2019-ci-image.qcow2

This is what all the other Ceph Windows jobs use.

Do you know if I can script that in the job ?

Reply:
I just created an SSH key that will let the job push the qcow2 image.

You'll need to

1. Add the `credentials-binding` [plugin](https://docs.openstack.org/infra/jenkins-job-builder/wrappers.html#wrappers.credentials-binding) to this job and use the `CEPH_WINDOWS_FILEDUMP_SSH_KEY` credential ID.  So:
```
- credentials-binding:
    - ssh-user-private-key:
       credential-id: CEPH_WINDOWS_FILEDUMP_SSH_KEY
       key-file-variable: FILEDUMP_SSH_KEY
       username-variable: FILEDUMP_USER
```
2. rsync it like `rsync -azvP --rsync-path='sudo rsync' new.qcow2 $FILEDUMP_USER@drop.front.sepia.ceph.com:/ceph/filedump.ceph.com/windows/`
Maybe write the SSH key to a temporary file and add `-e "ssh -i $WORKSPACE/tmp_keyfile"`

You can also use that key to SSH and write the new sha256 sum file.

Reply:
:+1:

Reply:
I've updated the job to use the new set of SSH credentials.

I run a test, and the image was successfully updated at: https://filedump.ceph.com/windows/.